pkg/client/fake/client.go

```diff
@@ -972,6 +974,19 @@ func copyStatusFrom(old, new runtime.Object) error {
 	return nil
 }
 
+// copyFrom copies from old into new
+func copyFrom(old, new runtime.Object) error {
+	oldMapStringAny, err := toMapStringAny(old)
```

Comment:
The original reason for not doing this anymore is that it might leave stale data in the target. I am confused as to why this isn't a problem anymore. Maybe we just didn't add sufficient test coverage?

Reply:
This isn't changing existing behavior. No tests were modified or removed. What it does is handle a mix of  types and unstructured objects gracefully, instead of panicing.

What stale data is left behind?

Reply:
What @alvaroaleman is referring to I believe is that before this function was added, we removed a function similar to this where we were doing a `copyFrom` because of the issues with that function leaving stale data around. 
https://github.com/kubernetes-sigs/controller-runtime/pull/2479

https://github.com/kubernetes-sigs/controller-runtime/pull/2479#discussion_r1320627434

If this being added doesn't impact what is already there (covers previous tests and your new test case) and handles the new case to fix the regression, the new test coverage should be sufficient to have this fix be in and resolve the issue with unstructured stuff.  

I believe we have unstructured.Unstructured tests cases and if not being caught in testing, it begs the question how did those tests pass, if this problem still presented itself. 

_IANAM, (I am not a maintainer)_

Reply:
the previous tests and code handled both `typed->typed` and `unstructured->unstructured`, but triggered a panic with `typed->unstructured`. The mix can happen when the GVK is registered in the scheme, so the resource in the tracker is always converted to the typed form, and then the user performs a status update using an unstructured object.

It's a bit esoteric as most users will favor a typed resource when available, but we hit it.

Reply:
Ah, we also added code to [zero the target](https://github.com/kubernetes-sigs/controller-runtime/blob/6860cfb3d210ac028dc5d744e7824a99c8841f07/pkg/client/fake/client.go#L1015), that is why it works. Thanks for the patch and sorry for the review delay